datafusion-examples/examples/simple_udtf.rs

```diff
@@ -0,0 +1,224 @@
+use datafusion::physical_plan::memory::MemoryExec;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::physical_plan::streaming::PartitionStream;
+use datafusion::physical_plan::{collect, ExecutionPlan};
+use datafusion::prelude::SessionContext;
+use datafusion_common::{DFSchema, ScalarValue};
+use datafusion_expr::{EmptyRelation, Expr, LogicalPlan, Projection, TableType};
+use std::fs::File;
+use std::io::Seek;
+use std::path::Path;
+use std::sync::Arc;
+
+// To define your own table function, you only need to do the following 3 things:
+// 1. Implement your own TableProvider
+// 2. Implement your own TableFunctionImpl and return your TableProvider
+// 3. Register the function using ctx.register_udtf
+
+/// This example demonstrates how to register a TableFunction
+#[tokio::main]
+async fn main() -> Result<()> {
+    // create local execution context
+    let ctx = SessionContext::new();
+
+    ctx.register_udtf("read_csv", Arc::new(LocalCsvTableFunc {}));
+    ctx.register_udtf("read_csv_stream", Arc::new(LocalStreamCsvTable {}));
+
+    let testdata = datafusion::test_util::arrow_test_data();
+    let csv_file = format!("{testdata}/csv/aggregate_test_100.csv");
+
+    // run it with println now()
+    let df = ctx
+        .sql(format!("SELECT * FROM read_csv('{csv_file}', now());").as_str())
+        .await?;
+    df.show().await?;
+
+    // just run
+    let df = ctx
+        .sql(format!("SELECT * FROM read_csv('{csv_file}');").as_str())
+        .await?;
+    df.show().await?;
+
+    // stream csv table
+    let df2 = ctx
+        .sql(format!("SELECT * FROM read_csv_stream('{csv_file}');").as_str())
+        .await?;
+    df2.show().await?;
+
+    Ok(())
+}
+
+// Option1: (full implmentation of a TableProvider)
+struct LocalCsvTable {
+    schema: SchemaRef,
+    exprs: Vec<Expr>,
+    batches: Vec<RecordBatch>,
+}
+
+#[async_trait]
+impl TableProvider for LocalCsvTable {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        _filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        if !self.exprs.is_empty() {
+            self.interpreter_expr(state).await?;
+        }
+        Ok(Arc::new(MemoryExec::try_new(
+            &[self.batches.clone()],
+            TableProvider::schema(self),
+            projection.cloned(),
+        )?))
+    }
+}
+
+impl LocalCsvTable {
+    // TODO(veeupup): maybe we can make interpreter Expr this more simpler for users
+    // TODO(veeupup): maybe we can support more type of exprs
+    async fn interpreter_expr(&self, state: &SessionState) -> Result<()> {
+        use datafusion::logical_expr::expr_rewriter::normalize_col;
```

Comment:
Perhaps for this example, we implement something simpler like a projecton -- so like  `read_csv('filename', 2)` would read only the first 2 rows from the file? 

That would be more inline with my expected usecase of changing the output of the table based on the input arguments.